tests/frontend/test_io.py

```diff
@@ -0,0 +1,543 @@
+        name test_program
+        version 1.0
+        """
+
+        bb = blackbird.loads(bb_script)
+
+        with pytest.raises(ValueError, match="contains no quantum operations"):
+            io.to_program(bb)
+
+    def test_gate_not_defined(self):
+        """Test empty program raises error"""
+
+        bb_script = """\
+        name test_program
+        version 1.0
+
+        Toffoli | [0, 1]
+        """
+
+        bb = blackbird.loads(bb_script)
+
+        with pytest.raises(NameError, match="operation Toffoli not defined"):
+            io.to_program(bb)
+
+    def test_metadata(self):
+        """Test metadata converts"""
+
+        bb_script = """\
+        name test_program
+        version 1.0
+        Vac | 0
+        """
+
+        bb = blackbird.loads(bb_script)
+        prog = io.to_program(bb)
+
+        assert prog.name == bb.name
+
+    def test_gate_no_arg(self):
+        """Test gate with no argument converts"""
+
+        bb_script = """\
+        name test_program
+        version 1.0
+        Vac | 0
+        """
+
+        bb = blackbird.loads(bb_script)
+        prog = io.to_program(bb)
+
+        assert len(prog) == 1
+        assert prog.circuit[0].op.__class__.__name__ == "Vacuum"
+        assert prog.circuit[0].reg[0].ind == 0
+
+    def test_gate_arg(self):
+        """Test gate with arguments converts"""
+
+        bb_script = """\
+        name test_program
+        version 1.0
+        Sgate(0.54, 0.12) | 0
+        """
+
+        bb = blackbird.loads(bb_script)
+        prog = io.to_program(bb)
+
+        assert len(prog) == 1
+        assert prog.circuit[0].op.__class__.__name__ == "Sgate"
+        assert prog.circuit[0].op.p[0].x == 0.54
+        assert prog.circuit[0].op.p[1].x == 0.12
+        assert prog.circuit[0].reg[0].ind == 0
+
+    def test_gate_kwarg(self):
+        """Test gate with keyword argument converts"""
+
+        bb_script = """\
+        name test_program
+        version 1.0
+
+        Dgate(a=0.54) | 0
+        """
+
+        bb = blackbird.loads(bb_script)
+        prog = io.to_program(bb)
+
+        assert len(prog) == 1
+        assert prog.circuit[0].op.__class__.__name__ == "Dgate"
+        assert prog.circuit[0].op.p[0].x == 0.54
+        assert prog.circuit[0].reg[0].ind == 0
+
+    def test_gate_multimode(self):
+        """Test gate with keyword argument converts"""
+
+        bb_script = """\
+        name test_program
+        version 1.0
+
+        BSgate(theta=0.54, phi=pi) | [0, 2]
+        """
+
+        bb = blackbird.loads(bb_script)
+        prog = io.to_program(bb)
+
+        assert len(prog) == 1
+        assert prog.circuit[0].op.__class__.__name__ == "BSgate"
+        assert prog.circuit[0].op.p[0].x == 0.54
+        assert prog.circuit[0].op.p[1].x == np.pi
+        assert prog.circuit[0].reg[0].ind == 0
+        assert prog.circuit[0].reg[1].ind == 2
+
+    def test_compilation(self):
+        """Test automatic compilation using the target keyword"""
```

Comment:
Not sure we want circuits to be implicitly compiled

Reply:
Resolving this comment since we have moved the discussion below.